Fixed #16501 -- Added a unicode parameter to models.SlugField and forms....

kutenai opened the pull request:
This is a rebase and update of pull request #3729 

Added unicode support to URLify and prepopulate scripts.
Matched unicode regex with current one.

Fixed an indentation error

Renamed unicode parameter to allow_unicode

timgraham commented on django/contrib/admin/options.py:
```diff
@@ -111,7 +111,7 @@ class BaseModelAdmin(six.with_metaclass(forms.MediaDefiningClass)):
     @classmethod
     def check(cls, model, **kwargs):
-        return cls.checks_class().check(cls, model, **kwargs)
+            return cls.checks_class().check(cls, model, **kwargs)
```
indent inadvertently modified

timgraham commented on django/contrib/admin/options.py:
```diff
@@ -1744,6 +1745,9 @@ def __init__(self, parent_model, admin_site):
     def media(self):
         extra = '' if settings.DEBUG else '.min'
         js = ['jquery%s.js' % extra, 'jquery.init.js', 'inlines%s.js' % extra]
+        # TODO: Was removed in master, modified in patch.
```
We can remove this - it's always included above.

kutenai replied:
I do not see where this is added above. The previous line does not add it, so I'm wondering if there is something else missing.

timgraham replied:
The idea is that `InlineModelAdmin` is always used with a `ModelAdmin`, so we don't need to include the files again.

timgraham commented on django/contrib/admin/static/admin/js/prepopulate.js:
```diff
@@ -1,11 +1,12 @@
 (function($) {
-    $.fn.prepopulate = function(dependencies, maxLength) {
+    $.fn.prepopulate = function(dependencies, maxLength, unicode) {
```
let's call the new variable `allowUnicode`

timgraham commented on django/db/models/fields/__init__.py:
```diff
@@ -2112,6 +2112,9 @@ def __init__(self, *args, **kwargs):
         # Set db_index=True unless it's been set manually.
         if 'db_index' not in kwargs:
             kwargs['db_index'] = True
+        self.unicode = kwargs.pop('unicode', False)
```
and let's call it `allow_unicode` on the Python side.

timgraham commented on docs/ref/forms/fields.txt:
```diff
@@ -847,6 +847,17 @@ For each field, we describe the default widget used if you don't specify
+       A boolean instructing the field to accept unicode letters. Defaults to
+       ``False``.
+
+   .. versionchanged:: 1.8
```
1.8 should be 1.9 now

timgraham commented on docs/ref/validators.txt:
```diff
@@ -183,6 +183,15 @@ to, or in lieu of custom ``field.clean()`` methods.
 
+``validate_unicode_slug``
+-------------------------
+.. data:: validate_slug
```
please add a blank line between ------ and .. data ...

timgraham commented on docs/releases/1.8.txt:
```diff
@@ -358,6 +358,9 @@ Forms
 * You can now pass a callable that returns an iterable of choices when
   instantiating a :class:`~django.forms.ChoiceField`.
 
+* :class:`~django.forms.SlugField` now accepts a
```
please move to 1.9 release notes

timgraham commented on tests/model_fields/tests.py:
```diff
@@ -847,14 +856,14 @@ def test_IntegerField(self):
-        self.assertIsInstance(
-            IPAddressField().get_prep_value(lazy_func()),
-            six.text_type)
+            lazy_func = lazy(lambda: '127.0.0.1', six.text_type)
```
inadvertent indent change

kutenai commented:
I've obviously messed this up, and I'm working on resolving the issues. When I changed the attribute names, it caused a ripple effect that I"m working through.

kutenai commented:
I believe I have addressed your issues now tim. I've also verified that the tests, docs, etc. run clean.

I had to add an unrelated fix to this file:
tests/gis_tests/distapp/tests.py

It was importing Point, but I don't have GIS, so for me it was failing. It was importing Point, which is not available if HAS_GIS is False.

This patch could use a final rebase when it is fully clean.

timgraham commented on django/contrib/admin/options.py:
```diff
@@ -1744,8 +1745,6 @@ def __init__(self, parent_model, admin_site):
     def media(self):
         extra = '' if settings.DEBUG else '.min'
         js = ['jquery%s.js' % extra, 'jquery.init.js', 'inlines%s.js' % extra]
-        if self.filter_vertical or self.filter_horizontal:
```
don't think this should be removed as part of this patch?

timgraham commented on django/contrib/admin/templates/admin/prepopulated_fields_js.html:
```diff
@@ -8,7 +8,8 @@
         dependency_list: [],
-        maxLength: {{ field.field.field.max_length|default_if_none:"50"|unlocalize }}
+        maxLength: {{ field.field.field.max_length|default_if_none:"50"|unlocalize }},
+        unicode: {{ field.field.field.unicode|lower }}
```
unicode -> allow_unicode, right? maybe we need to add a test for this...

timgraham commented on django/contrib/staticfiles/handlers.py:
```diff
@@ -1,6 +1,7 @@
 from django.conf import settings
 from django.contrib.staticfiles import utils
 from django.contrib.staticfiles.views import serve
+from django.core import urlresolvers
```
Think the changes to this file are inadvertent?

kutenai replied:
I added it because urlresolvers was not imported, but it was used.

timgraham replied:
I meant, what is the purpose of the changes? They don't appear to be part of the original patch. Are they related to supporting unicode in `SlugField`?

timgraham commented on django/db/models/fields/__init__.py:
```diff
@@ -2122,13 +2125,15 @@ def deconstruct(self):
             kwargs['db_index'] = False
         else:
             del kwargs['db_index']
+        if self.allow_unicode is not False:
```
I think `if not self.allow_unicode` would be fine

timgraham commented on tests/forms_tests/tests/test_fields.py:
```diff
@@ -1509,6 +1509,18 @@ def test_slugfield_normalization(self):
+        self.assertEqual(f.clean('  你-好  '), '你-好')
+        self.assertEqual(f.clean('ıçğüş'), 'ıçğüş')
+        self.assertEqual(f.clean('foo-ıç-bar'), 'foo-ıç-bar')
+        # Suggested in pull request, but questions remain
```
can you elaborate on what the issue is here?

timgraham commented on tests/gis_tests/distapp/tests.py:
```diff
@@ -1,6 +1,6 @@
 from __future__ import unicode_literals
 
-from django.contrib.gis.geos import HAS_GEOS, Point
```
Thanks, I committed this change separately so you can rebase your branch so it doesn't appear anymore.

timgraham commented on django/db/models/fields/__init__.py:
```diff
@@ -374,7 +374,7 @@ def deconstruct(self):
         following types (including inner values of collection types):
 
-         * None, bool, str, unicode, int, long, float, complex, set, frozenset, list, tuple, dict
+         * None, bool, str, allow_unicode, int, long, float, complex, set, frozenset, list, tuple, dict
```
Revert change on this line (caused by find/replace, I guess).

timgraham commented:
@kutenai, do you have time/interest in completing this soon? I could offer it as a sprint task at DjangoconEU if not. Thanks!

kutenai commented:
Actually, Yes.
I've been out of commission for a while due to family issues and work 
issues. A perfect storm if you will.

I'll get back on this very soon.. sorry for the delay



> Tim Graham <mailto:notifications@github.com>
> June 2, 2015 at 8:12 AM
>
> @kutenai <https://github.com/kutenai>, do you have time/interest in 
> completing this soon? I could offer it as a sprint task at DjangoconEU 
> if not. Thanks!
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/django/django/pull/4518#issuecomment-107968858>.
>

kutenai commented:
All of the tests work for me locally, but fail here. I think I need a little help understanding the errors in the build.

kutenai commented:
There is an error in master build. I just rebased against master so that is the source of the above error.

timgraham commented:
Rebased & updated in https://github.com/django/django/pull/5010. Thanks!
